File: C++/31. Next Permutation.cpp
void nextPermutation(vector<int>& nums) 
{ 
    int i = nums.size()-1 ; 
    while(--i >= 0)
        if(nums[i] < nums[i+1])
            break ; 

    int j ; 
    if(i >= 0)
    {
        j = i+1 ;
        while(j < nums.size() and nums[i] < nums[j])
            j++ ;
        j-- ; 

        swap(nums[i], nums[j]);
    }

    j = nums.size()-1;
    i++; 
    while( i < j )
        swap(nums[i++], nums[j--]);
}

/*

3 2 1

1 5 8 4 7 6 5 3 1 // finding the decreasing point
1 5 8 5 7 6 4 3 1 // swap them with next greater 
1 5 8 5 1 3 4 6 7 // reverse the reamining 

*/

File: Rust/31. Next Permutation.rs
pub fn next_permutation(nums: &mut Vec<i32>) {
    let len = nums.len();
    let mut break_point = usize::MAX;

    if len < 2 {
        return;
    }

    // Find the breaking point
    /* Movement when highest point go down to lower */
    for i in (0..=len - 2).rev() {
        if nums[i] < nums[i + 1] {
            break_point = i;
            break;
        }
    }

    // if there is no breaking point then this must be the last perumutation simply return the 1'st perumutation
    if break_point == usize::MAX {
        nums.reverse();
        return;
    }

    // swap the break_point with higher than this from reverse order check
    for i in (break_point + 1..=len - 1).rev() {
        if nums[i] > nums[break_point] {
            nums.swap(break_point, i);
            break;
        }
    }

    // sort the rest of the array or reverse it since we know it is going increasing order
    let remaining_vec = &mut nums[break_point + 1..len];
    remaining_vec.reverse();
}
